include/oneapi/tbb/detail/_task_handle.h

```diff
@@ -136,32 +136,52 @@ class task_dynamic_state {
 };
 #endif // __TBB_PREVIEW_TASK_GROUP_EXTENSIONS
 
+template <typename DerivedType>
+void finalize_base(task_handle_task* p, d1::small_object_allocator& alloc, const d1::execution_data* ed) {
+    if (ed) {
+        alloc.delete_object(static_cast<DerivedType*>(p), *ed);
+    } else {
+        alloc.delete_object(static_cast<DerivedType*>(p));
+    }
+}
+
 class task_handle_task : public d1::task {
-    std::uint64_t m_version_and_traits{};
+    // The pointer to the instantiation of finalize_base with the concrete derived type class for correct finalization
+    // Reuse the first std::uint64_t field (was m_version_and_traits) to preserve backward compatibility
+    using finalize_func_type = void (*)(task_handle_task*, d1::small_object_allocator&, const d1::execution_data*);
+    std::uint64_t m_finalize_func;
+    static_assert(sizeof(finalize_func_type) <= sizeof(m_finalize_func), "Cannot fit finalize pointer into std::uint64_t");
```

Comment:
Why not like this?
```suggestion
    static_assert(sizeof(finalize_func_type) <= sizeof(std::uint64_t), "Cannot fit finalize pointer into std::uint64_t");
    alignas(std::uint64_t) finalize_func_type m_finalize_func;
```

And I would also use destroy/delete in the names here.

Reply:
Forcing the alignment requirement on the first field alone is not sufficient. If `sizeof(pointer)` is less than `sizeof(std::uint64_t)`, the overall size of structure and the offsets of subsequent fields will be affected.
Consider a 32-bit build:
* `sizeof(pointer)` is `4`.
* `alignof(std::uint64_t)` is `4`.

In this case:
* The pointer to the destroy function occupies offset `0-4`.
* The next field (also a pointer) occupies offset `4-8`.

In the old layout:
* `m_version_and_traits` was of type `std::uint64_t` (size = 8, alignment = 4).
* It occupied offset `0-8`, and the next pointer - the offset `8-12`.

This difference in layout breaks the ABI compatibility.
See [this example](https://godbolt.org/z/EnqPfoc31) for more details.

If we don't want to keep the type of the first field as `std::uint64_t `and use `reinterpret_cast` to access the destroy function, we can manually 
insert padding to preserve the original layout:

```cpp
struct task_handle_task {
    alignas(std::uint64_t) destroy_func_type m_destroy_func;
    char padding[sizeof(std::uint64_t) - sizeof(destroy_func_type)];
    wait_tree_vertex_interface* m_wait_tree_vertex;
};
```

However, I prefer to keep the type as `std::uint64_t` and simply add a comment explainng the rationale.

Reply:
Thanks for the explanation; I should have noticed that myself.
You could use an unnamed union of `std::uint64_t` and the pointer - though this would complicate initialization a bit. Anyway, I do not mind `reinterpret_cast` either.